lib/auto_attach.py

```diff
@@ -50,21 +54,41 @@ def check_cloudinit_userdata_for_ua_info():
 
 
 def main(cfg: UAConfig):
-    if not check_cloudinit_userdata_for_ua_info():
-        if actions.should_disable_auto_attach(cfg):
-            return
-        try:
-            full_auto_attach(FullAutoAttachOptions())
-        except AlreadyAttachedError as e:
-            logging.info(e.msg)
-    else:
-        auto_attach_msg = (
+    if actions.should_disable_auto_attach(cfg):
+        return
+
+    if check_cloudinit_userdata_for_ua_info():
+        logging.info("cloud-init userdata has ubuntu-advantage key.")
+        logging.info(
             "Skipping auto-attach and deferring to cloud-init "
             "to setup and configure auto-attach"
         )
+        return
 
-        logging.info("cloud-init userdata has ubuntu-advantage key.")
-        logging.info(auto_attach_msg)
+    system.write_file(
+        AUTO_ATTACH_STATUS_MOTD_FILE, messages.AUTO_ATTACH_RUNNING
+    )
+    try:
+        full_auto_attach(FullAutoAttachOptions())
+    except AlreadyAttachedError as e:
+        logging.info(e.msg)
+    except Exception as e:
+        logging.error(e)
+        system.remove_file(AUTO_ATTACH_STATUS_MOTD_FILE)
+        logging.info("creating flag file to trigger retries")
+        system.create_file(retry_auto_attach.FLAG_FILE_PATH)
```

Comment:
I am wondering if we should only retry on the following exceptions:

* `ConnectivityError`
* `ContractAPIError`
* `LockHeldError`

The other seem that will not benefit much by triggering the retry job

Reply:
I think I'll do the NonAutoAttachImageError refactor in a different PR, to be merged before this one

Reply:
I think now after all the refactoring, we might reduce the list of exceptions to just:

* ConnectivityError
* ContractAPIError
* UrlError
* LockHeldError
* InvalidProImage


I still don't know if we should try on `UserFacingError`. We could be masking a real issue for a month and a cannot think of a case where a retry might solve it. But I know retrying on that should be useful for solving those unpredictable situations, I am just still not sure if it can do more harm than good here